components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,340 @@
+                                        Map<String, Object> eventContext, ActionExecutionRequest actionRequest,
+                                        String actionId, String apiEndpoint,
+                                        AuthMethods.AuthMethod authenticationMethod,
+                                        ActionExecutionResponseProcessor actionExecutionResponseProcessor)
+            throws ActionExecutionException {
+
+        if (log.isDebugEnabled()) {
+            // todo: add to diagnostic logs
+            logSuccessResponse(successResponse, actionType, actionId, apiEndpoint, authenticationMethod);
+        }
+
+        List<PerformableOperation> allowedPerformableOperations =
+                validatePerformableOperations(actionRequest, successResponse);
+        ActionInvocationSuccessResponse.Builder successResponseBuilder =
+                new ActionInvocationSuccessResponse.Builder().setOperations(allowedPerformableOperations);
+        actionExecutionResponseProcessor.processSuccessResponse(actionType, eventContext, actionRequest.getEvent(),
+                successResponseBuilder.build());
+    }
+
+    private void processErrorResponse(ActionInvocationErrorResponse errorResponse, ActionType actionType,
+                                      Map<String, Object> eventContext, ActionExecutionRequest actionRequest,
+                                      String actionId, String apiEndpoint,
+                                      AuthMethods.AuthMethod authenticationMethod,
+                                      ActionExecutionResponseProcessor actionExecutionResponseProcessor)
+            throws ActionExecutionException {
+
+        if (log.isDebugEnabled()) {
+            // todo: add to diagnostic logs
+            logErrorResponse(errorResponse, actionType, actionId, apiEndpoint, authenticationMethod);
+        }
+
+        actionExecutionResponseProcessor.processErrorResponse(actionType, eventContext, actionRequest.getEvent(),
+                errorResponse);
+    }
+
+    private void logSuccessResponse(ActionInvocationSuccessResponse successResponse, ActionType actionType,
+                                    String actionId, String apiEndpoint, AuthMethods.AuthMethod authenticationMethod) {
+
+        try {
+            String responseBody = serializeSuccessResponse(successResponse);
+            log.debug(String.format(
+                    "Received success response from API: %s for action type: %s action id: %s with authentication: %s. Response: %s",
+                    apiEndpoint,
+                    actionType,
+                    actionId,
+                    Optional.ofNullable(authenticationMethod).map(AuthMethods.AuthMethod::getAuthType)
+                            .orElse("NONE"),
+                    responseBody));
+        } catch (JsonProcessingException e) {
+            log.error("Error occurred while deserializing the success response for action: " +
```

Comment:
Shall we follow the String.format() method in here as well.

Reply:
concatenation is faster than string format. String format brings readability only. So for debug logs it makes sense

Reply:
I followed the concatenation through out. So let me keep that for consistency